frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/ApplicationMaster.java

```diff
@@ -321,41 +322,36 @@ private void stopForInternalUnKnownError(String customizedDiagnostics) {
     stop(new StopStatus(ExitStatusKey.AM_INTERNAL_UNKNOWN_ERROR.toInt(), true, diagnostics));
   }
 
-  // Principle to setup ContainerRequest for a Task:
-  // 1. Exactly match the Task's Requirement
-  //    -> Keeps Waiting Allocation, i.e. containerRequestTimeoutSec = -1
-  // 2. Too Relax for the Task's Requirement
-  //    -> Reject Allocation and Re-Request, See testContainer
-  // 3. Too Strict for the Task's Requirement
-  //    -> Timeout Request and Re-Request, i.e. containerRequestTimeoutSec != -1
   private ContainerRequest setupContainerRequest(TaskStatus taskStatus) throws Exception {
     String taskRoleName = taskStatus.getTaskRoleName();
     Priority requestPriority = statusManager.getNextContainerRequestPriority();
     String requestNodeLabel = requestManager.getTaskPlatParams().get(taskRoleName).getTaskNodeLabel();
-    String requestNodeGpuType = requestManager.getTaskPlatParams().get(taskRoleName).getTaskNodeGpuType();
+
     ResourceDescriptor requestResource = requestManager.getTaskResources().get(taskRoleName);
     ResourceDescriptor maxResource = conf.getMaxResource();
 
     if (!ResourceDescriptor.fitsIn(requestResource, maxResource)) {
       LOGGER.logWarning(
           "Request Resource does not fit in the Max Resource configured in current cluster, " +
-              "request may be fail or never got satisfied: " +
+              "request may fail or never get satisfied: " +
               "Request Resource: [%s], Max Resource: [%s]",
           requestResource, maxResource);
     }
 
-    if (requestResource.getGpuNumber() > 0) {
-      updateNodeReports(yarnClient.getNodeReports(NodeState.RUNNING));
+    updateNodeReports(yarnClient.getNodeReports(NodeState.RUNNING));
+    SelectionResult selectionResult = selectionManager.select(requestResource, taskRoleName);
 
-      SelectionResult selectionResult = selectionManager.select(requestResource, requestNodeLabel, requestNodeGpuType);
-      if (selectionResult != null) {
-        ResourceDescriptor optimizedRequestResource = YamlUtils.deepCopy(requestResource, ResourceDescriptor.class);
-        optimizedRequestResource.setGpuAttribute(selectionResult.getGpuAttribute());
-        return HadoopUtils.toContainerRequest(optimizedRequestResource, requestPriority, null, selectionResult.getNodeHost());
-      }
+    ResourceDescriptor optimizedRequestResource = selectionResult.getOptimizedResource();
+    if (selectionResult.getNodeHosts().size() <= 0) {
+      return HadoopUtils.toContainerRequest(optimizedRequestResource, requestPriority, requestNodeLabel, null);
     }
 
-    return HadoopUtils.toContainerRequest(requestResource, requestPriority, requestNodeLabel, null);
+    // Random pick a host from the result set to avoid conflicted requests from concurrent container requests
+    // from different jobs
+    int random = new Random().nextInt(selectionResult.getNodeHosts().size());
+    String candidateNode = selectionResult.getNodeHosts().get(random);
```

Comment:
Use CommonUtils.getRandomNumber

Reply:
format

---
In reply to: [174685949](https://github.com/Microsoft/pai/pull/318#discussion_r174685949) [](ancestors = 174685949)